### PR TITLE
Rebalance when busy, make min free disk absolute

### PR DESCRIPTION
**Pull request by tedpearson**
These are some quick fixes to make rebalancing keep up better and clean up some code.

- change disk free space to force rebalance or not allow new tasks from percent to absolute numbers
- change host score equation to not include cluster max values, each host’s score is independent
- move some methods inside of hostscore: getDiskUsedPercent, getAvailDiskBytes

### CODE REVIEW

File: hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancerConfig.java [hunk 1]
```diff
@@ -72,10 +72,10 @@
     // Track the last time a job autobalance was done
     private long lastJobAutobalanceTime = 0L;
 
-    // If you're at more than 80% disk capacity, you don't get assigned any new tasks or replicas
-    private double minDiskPercentAvailToReceiveNewTasks = Double.parseDouble(Parameter.value("spawnbalance.min.disk.percent.avail.newtasks", ".2"));
-    // If you're at more than 90% disk capacity, push tasks off before running additional tasks
-    private double minDiskPercentAvailToRunJobs = Double.parseDouble(Parameter.value("spawnbalance.min.disk.percent.avail.runtasks", ".1"));
+    // If you have less than 700GB free disk space, push tasks off before running additional tasks
```
Comment (yuesong) [inline hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancerConfig.java]: Comment should not be the same as that of the parameter below?

File: hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java [hunk 2]
```diff
@@ -1256,26 +1243,18 @@ protected void updateAggregateStatistics(List<HostState> hosts) {
...
-        score += activeTaskWeight * Math.pow(meanActive / clusterMaxMeanActive, 2.5);
-        score += diskUsedWeight * Math.pow(usedDiskPercent / clusterMaxDiskUsed, 2.5);
+        double exponent = 2.5;
+        double score = activeTaskWeight * Math.pow(meanActive, exponent);
+        double diskPercentPowered = Math.pow(diskUsedPercentModified, exponent);
+        score += diskUsedWeight * diskPercentPowered;
         // If host is very full, make sure to give the host a big score
-        score = Math.max(score, (activeTaskWeight + diskUsedWeight) * usedDiskPercent);
-        return new HostScore(meanActive, usedDiskPercent, score);
+        score = Math.max(score, (activeTaskWeight + diskUsedWeight) * diskPercentPowered);
```
Comment (yuesong) [inline hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java]: Old code uses `usedDiskPercent` while new code uses `diskPercentPowered`. Is that intentional?

File: hydra-main/src/main/java/com/addthis/hydra/job/mq/HostState.java [hunk 3]
```diff
@@ -391,6 +391,24 @@ public void setMax(HostCapacity max) {
...
+
+    public double getDiskUsedPercent() {
+        return getDiskUsedPercentModified(0);
+    }
+
+    public double getDiskUsedPercentModified(long reduce) {
+        if ((max == null) || (used == null) || (max.getDisk() <= 0)) {
+            return 0;
+        }
+        return (double) (used.getDisk() / (max.getDisk() - reduce));
```
Comment (yuesong) [inline hydra-main/src/main/java/com/addthis/hydra/job/mq/HostState.java]: Should check return value to make sure it's in the range of [0, 1]. Also add javadoc here and/or comment where it is called to explain usage/reasoning

File: hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancerConfig.java (hunk 1 above)
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancerConfig.java]: I was hoping I could sneak this through and fix it later 😛

File: hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java (hunk 2 above)
Comment (tedpearson) [inline hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java]: Old code was wrong?

**Comment by tedpearson** [comment]
Comments addressed

File: hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java (hunk 2 above)
Comment (yuesong) [inline hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/SpawnBalancer.java]: I still don't quite understand how it gives host with fuller disk a higher score.

The default configuration is: activeTaskWeight=30 and diskUsedWeight=70. meanActive is about 2 and doesn't seem to change a lot over time, according to ganglia metric. Now let's say disk is full, so diskUsedPercentModified=1

The original formula (before the Math.max call) produces: score=30*pow(2, 2.5) + 70*pow(1, 2.5)=170+70=240
The alternative formula (second parameter in the Math.max call) produces: score=(30+70)*pow(1, 2.5)=100

The original score still wins.

The thing is: disk percent is <=1 while mean active seems to be around 2. With the exponential, mean active is almost always going to be the dominant factor.

Where am I wrong in this?

**Comment by tedpearson** [comment]
You are wrong in that `host.getMeanActiveTasks()` returns a value from 0 to 1, and would never return 2. See Minion.java line 660.

**Comment by yuesong** [comment]
Good enough.

**Pull request merged**